### PR TITLE
test(hc): Exempt test setup from silo mode txn validation

### PR DESCRIPTION
**Pull request by RyanSkonnord**
~~Add two Pytest fixtures to swap into monolith mode before the built-in `django_db_setup` fixture and back after. This should prevent a false error from `validate_transaction_using_for_silo_mode` when we eventually switch the default silo mode for tests from monolith to region.~~

Now a different approach. See explanation in docstring.

### CODE REVIEW

**Comment by RyanSkonnord** [comment]
Depends on https://github.com/getsentry/sentry/pull/63916.

Fixes errors observed in https://github.com/getsentry/sentry/pull/63914 when switching to region mode by default.

**Comment by RyanSkonnord** [comment]
To do: check whether the mode swap can be done through `SingleProcessSiloModeState` rather than by mutating Django settings.

**Review by dashed — APPROVED** [review]

**Review by dashed — APPROVED** [review]

**Comment by RyanSkonnord** [comment]
Downgrading to draft. I thought I had this working yesterday but the state of the CI runs from https://github.com/getsentry/sentry/pull/63914 suggests it's not. Something about the sequence of the fixtures.

File: src/sentry/silo/patches/silo_aware_transaction_patch.py [hunk 1]
```diff
@@ -46,13 +48,52 @@ def siloed_on_commit(func: Callable[..., Any], using: Optional[str] = None) -> N
...
 
+def is_in_test_case_body() -> bool:
+    """Determine whether the current execution stack is in a test case body.
+
+    This is a best-effort, potentially brittle implementation that depends on private
+    behavior of the current Pytest implementation. We can't necessarily rely on
+    underscore-prefixed method names being used in a stable way.
+
+    Are you landing here because test cases regressed mysteriously after a Pytest
+    upgrade? Check the list of frames and add tweak the condition logic to make this
```
Comment (GabeVillalobos) [inline src/sentry/silo/patches/silo_aware_transaction_patch.py:59]: Minor typo here "and add tweak the..."

**Review by GabeVillalobos — APPROVED** [review]
Wow, this is a hell of a kludge but gets the job done. Nice work!

File: src/sentry/silo/patches/silo_aware_transaction_patch.py [hunk 2]
```diff
@@ -46,13 +48,52 @@ def siloed_on_commit(func: Callable[..., Any], using: Optional[str] = None) -> N
...
+
+
 def validate_transaction_using_for_silo_mode(using: Optional[str]):
     from sentry.models.outbox import ControlOutbox, RegionOutbox
     from sentry.silo import SiloMode
 
     if using is None:
         raise TransactionMissingDBException("'using' must be specified when creating a transaction")
 
+    if not is_in_test_case_body():
```
Comment (corps) [inline src/sentry/silo/patches/silo_aware_transaction_patch.py:94]: Nit: Explain at the call site here why this assumption is ok, but why it's not ok in other cases.  Basically, just some more context here would help pairs of 👁️

Comment (RyanSkonnord) [reply src/sentry/silo/patches/silo_aware_transaction_patch.py]: 👍 Done.

**Comment by RyanSkonnord** [comment]
And just for posterity, if anyone wants to attempt to salvage the partial, fixture-based solution mentioned in my comment, the magic invocation in `conftest.py` would look approximately like this:

```python
is_in_db_setup = False

@pytest.fixture(scope="session")
def change_state_before_db_setup():
    global is_in_db_setup
    is_in_db_setup = True

@pytest.fixture(scope="session")
def django_db_setup(change_state_before_db_setup, django_db_setup):
    global is_in_db_setup
    is_in_db_setup = False
```

**Pull request merged**